### PR TITLE
[WIP] [GSOC-2017] Import EventWaitHandle, AutoResetEvent, ManualResetEvent, Mutex and Semaphore

### PR DESCRIPTION
**Pull request by dampir**
In this post, I want to summarize what has been done during my participation in GSoC 2017.
I'm going to send the link to this post as my work product submission for final evaluation of GSoC 2017.

What has been done:
- [x] Added additional platform specific dll.sources files to handle the CoreRT approach when platform specific code is placed in its own file.
- [x] Ported all needed managed classes of CoreRT source code. It includes classes that must be ported to integrate Semaphore, Mutex and EventWaitHandle synchronization primitives into Mono.
- [x] The CoreRT runtime library System.Private.CoreLib.Native was integrated into Mono as a dynamic library.
- [x] Started a work on an adaptation of Mono runtime to handle of CoreRT's managed class instead of Mono's ones.

What should be done next:
- [ ] As was found, imported sync primitives are tightly connected with the Thread class from CoreFX project. The Thread class must be imported first, but it means that all System.Threading (the Tasks classes, etc) must be imported next. I tried to import them but the attempt was failed. Here is a link to the commits I did: [gsoc-import-thread-class](https://github.com/dampir/mono/commits/gsoc-import-thread-class).
- [ ] The work on an adaptation of Mono runtime should be done next.

I want to say big thanks for all people that helped me during GSoC 2017. I got a very valuable experience, especially in C# and Mono internals.

### CODE REVIEW

File: mcs/class/corlib/darwin_net_4_x_corlib.dll.exclude.sources [hunk 1]
```diff
@@ -0,0 +1 @@
+../../../external/corert/src/Common/src/Interop/Windows/mincore/Interop.MemAllocFree.cs
```
Comment (marek-safar) [inline mcs/class/corlib/darwin_net_4_x_corlib.dll.exclude.sources]: Move Windows/mincore/Interop.MemAllocFree.cs to win32_net_4_x_corlib.dll.sources and you can avoid all excludes.sources

File: mcs/class/System/darwin_net_4_x_System.dll.sources [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#include net_4_x_System.dll.sources
```
Comment (marek-safar) [inline mcs/class/System/darwin_net_4_x_System.dll.sources]: @luhenry is this necessary I think we had auto-fall back for this

**Comment by marek-safar** [comment]
build

File: mcs/class/corlib/darwin_net_4_x_corlib.dll.exclude.sources (hunk 1 above)
Comment (dampir) [inline mcs/class/corlib/darwin_net_4_x_corlib.dll.exclude.sources]: I'm almost sure that did it but there were some errors. Anyway, I'll try to change it again.

**Comment by marek-safar** [comment]
Could you investigate the test failures, some looks like related to your changes

**Comment by dampir** [comment]
Yes, I'll review them for now.

2017-08-15 14:57 GMT+07:00 Marek Safar <notifications@github.com>:

**Comment by dampir** [comment]
build

**Comment by dampir** [comment]
build

**Comment by dampir** [comment]
build

**Comment by dampir** [comment]
@marek-safar 
I investigated tests and made some changes. Could you recheck it?

**Comment by dampir** [comment]
build

**Pull request closed without merging**